Source/JavaScriptCore/dfg/DFGMovHintRemovalPhase.cpp

```diff
@@ -0,0 +1,155 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "DFGMovHintRemovalPhase.h"
+
+#if ENABLE(DFG_JIT)
+
+#include "DFGEpoch.h"
+#include "DFGForAllKills.h"
+#include "DFGGraph.h"
+#include "DFGInsertionSet.h"
+#include "DFGMayExit.h"
+#include "DFGPhase.h"
+#include "JSCJSValueInlines.h"
+#include "OperandsInlines.h"
+
+namespace JSC { namespace DFG {
+
+namespace {
+
+namespace DFGMovHintRemovalPhaseInternal {
+static constexpr bool verbose = false;
+}
+
+class MovHintRemovalPhase : public Phase {
+public:
+    MovHintRemovalPhase(Graph& graph)
+        : Phase(graph, "MovHint removal")
+        , m_insertionSet(graph)
+        , m_state(OperandsLike, graph.block(0)->variablesAtHead)
+        , m_changed(false)
+    {
+    }
+
+    bool run()
+    {
+        if (DFGMovHintRemovalPhaseInternal::verbose) {
+            dataLog("Graph before MovHint removal:\n");
+            m_graph.dump();
+        }
+
+        for (BasicBlock* block : m_graph.blocksInNaturalOrder())
+            handleBlock(block);
+
+        m_insertionSet.execute(m_graph.block(0));
+
+        return m_changed;
+    }
+
+private:
+    void handleBlock(BasicBlock* block)
+    {
+        if (DFGMovHintRemovalPhaseInternal::verbose)
+            dataLog("Handing block ", pointerDump(block), "\n");
+
+        // A MovHint is unnecessary if the local dies before it is used. We answer this question by
+        // maintaining the current exit epoch, and associating an epoch with each local. When a
+        // local dies, it gets the current exit epoch. If a MovHint occurs in the same epoch as its
+        // local, then it means there was no exit between the local's death and the MovHint - i.e.
+        // the MovHint is unnecessary.
+
+        Epoch currentEpoch = Epoch::first();
+
+        m_state.fill(Epoch());
+        m_graph.forAllLiveInBytecode(
+            block->terminal()->origin.forExit,
+            [&] (Operand reg) {
+                m_state.operand(reg) = currentEpoch;
+            });
+
+        if (DFGMovHintRemovalPhaseInternal::verbose)
+            dataLog("    Locals at ", block->terminal()->origin.forExit, ": ", m_state, "\n");
+
+        // Assume that blocks after us exit.
```

Comment:
What's `us exit`?

Reply:
This is a comment in the original RemoveMovHint phase file. We should change it to `we`.